File: bank-b/scripts/migrate.py
import asyncio
import os
import sys
from pathlib import Path

import asyncpg
from dotenv import load_dotenv

load_dotenv()

MIGRATIONS_DIR = Path(__file__).resolve().parent.parent / "migrations"

# Arbitrary key: serializes concurrent runners (e.g. two instances deploying at once).
ADVISORY_LOCK_KEY = 723532


async def migrate() -> None:
    connection = await asyncpg.connect(dsn=os.environ["DATABASE_URL"])
    try:
        await connection.execute("SELECT pg_advisory_lock($1)", ADVISORY_LOCK_KEY)
        await connection.execute("""
            CREATE TABLE IF NOT EXISTS schema_migrations (
                name       TEXT PRIMARY KEY,
                applied_at TIMESTAMPTZ NOT NULL DEFAULT now()
            )
        """)

        applied = {
            row["name"] for row in await connection.fetch("SELECT name FROM schema_migrations")
        }
        files = sorted(f for f in os.listdir(MIGRATIONS_DIR) if f.endswith(".sql"))

        for file in files:
            if file in applied:
                continue
            sql = (MIGRATIONS_DIR / file).read_text()
            async with connection.transaction():
                await connection.execute(sql)
                await connection.execute("INSERT INTO schema_migrations (name) VALUES ($1)", file)
            print(f"applied {file}")
    finally:
        await connection.execute("SELECT pg_advisory_unlock($1)", ADVISORY_LOCK_KEY)
        await connection.close()


if __name__ == "__main__":
    try:
        asyncio.run(migrate())
    except Exception as error:
        print(error, file=sys.stderr)
        sys.exit(1)

File: hub/scripts/migrate.ts
import 'dotenv/config';
import { readdir, readFile } from 'node:fs/promises';
import { join } from 'node:path';
import { Pool } from 'pg';

const MIGRATIONS_DIR = join(__dirname, '..', 'migrations');

// Arbitrary key: serializes concurrent runners (e.g. two instances deploying at once).
const ADVISORY_LOCK_KEY = 723533;

async function migrate(): Promise<void> {
  const pool = new Pool({ connectionString: process.env.DATABASE_URL });
  const client = await pool.connect();
  try {
    await client.query('SELECT pg_advisory_lock($1)', [ADVISORY_LOCK_KEY]);
    await client.query(`
      CREATE TABLE IF NOT EXISTS schema_migrations (
        name       TEXT PRIMARY KEY,
        applied_at TIMESTAMPTZ NOT NULL DEFAULT now()
      )
    `);

    const { rows } = await client.query<{ name: string }>('SELECT name FROM schema_migrations');
    const applied = new Set(rows.map((row) => row.name));
    const files = (await readdir(MIGRATIONS_DIR)).filter((file) => file.endsWith('.sql')).sort();

    for (const file of files) {
      if (applied.has(file)) continue;
      const sql = await readFile(join(MIGRATIONS_DIR, file), 'utf8');
      await client.query('BEGIN');
      try {
        await client.query(sql);
        await client.query('INSERT INTO schema_migrations (name) VALUES ($1)', [file]);
        await client.query('COMMIT');
        console.log(`applied ${file}`);
      } catch (error) {
        await client.query('ROLLBACK');
        throw new Error(`migration ${file} failed`, { cause: error });
      }
    }
  } finally {
    await client.query('SELECT pg_advisory_unlock($1)', [ADVISORY_LOCK_KEY]);
    client.release();
    await pool.end();
  }
}

migrate().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
